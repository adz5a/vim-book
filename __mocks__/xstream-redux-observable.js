module.exports = require("xstream-redux-observable/commonjs");
