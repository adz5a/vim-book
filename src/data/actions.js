export const ACTIONS = {
  tocLoaded: "toc-loaded",
  tocSaved: "toc-saved",
  errorTocSave: "error-toc-save",
  tocUnavailable: "toc-unavailable",
  chapterLoaded: "chapter-loaded",
  chapterUnavailable: "chapter-unavailable"
};


export const withType = type => action => action.type === type;
export const asType = type => data => ({
    type,
    data
});

export const makeReducer = (options, defaultState = {}) => {
    const reduce = ( state = defaultState, action ) => {
        const { type, data } = action;
        if ( options[type] ) {
            return options[type](state, data);
        } else return state;
    };
    return reduce;
};
