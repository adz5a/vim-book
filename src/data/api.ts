import xs, { Stream } from "xstream";


interface Options {
  repo: string;
  author: string;
  path?: string;
}

/*
 * https://developer.github.com/v3/repos/contents/
 */
interface ApiResponse {
  type: "file"| "dir";
  encoding: "base64" | null;
  size: number;
  name: string;
  path: string;
  content: string;
  sha: string;
  url: string;
  git_url: string;
  html_url: string;
  download_url: string;
  _links: {
    git: string;
    self: string;
    html: string;
  }
}

interface ApiPartialResponse {
  type: "file" | "dir";
  size: number;
  name: string;
  path: string;
  sha: string;
  url: string;
  git_url: string;
  html_url: string;
  download_url: string | null;
  _links: {
    self: string;
    git: string;
    html: string;
  }
}

const baseURL = "https://api.github.com";

export const makeURL = ( options: Options ) => {
  if (options.path !== undefined) {
    return `${baseURL}/repos/${options.author}/${options.repo}/contents/${options.path}`;
  } else {
    return `${baseURL}/repos/${options.author}/${options.repo}`;
  }
};

export const getContents = ( options: Options ): Promise<ApiResponse | ApiPartialResponse[]> => {
  return fetch(makeURL(options))
    .then( response => {
      return response.json()
        .then( data => {
          if ( response.status !== 200 ) {
            throw data;
          } else {
            return data;
          }
        });
    }) as Promise<ApiResponse>;
};

export const getContents$ = ( options: Options ): Stream<ApiResponse | ApiPartialResponse[]> => {
  return xs.fromPromise(getContents(options));
};
