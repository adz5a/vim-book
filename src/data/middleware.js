import { createMiddleware } from "xstream-redux-observable";
import { getContents$ } from "./api";
import { ACTIONS, withType, asType, makeReducer } from "./actions";
import xs from "xstream";
import delay from "xstream/extra/delay";

const author= "sjl";
const repo= "learnvimscriptthehardway";
const PouchDB = global.PouchDB;
const db = new PouchDB("book");
global.bookDB = db;

const fetchTOC = () => getContents$({
  author,
  repo,
  path: "chapters"
});

const updateCachedChapter = chapter => db.get(chapter.path)
  .then(doc => db.put({
    _id: chapter.path,
    _rev: doc._rev,
    chapter
  }), error => {
    if ( error.status === 404 ) {
      return db.put({
        _id: chapter.path,
        chapter
      });
    } else throw error;
  });

export const middleware = createMiddleware( action$ => {

  const loadTOC$ = action$.filter(withType(ACTIONS.loadTOC))
    .map(() => xs.fromPromise(db.get("toc")
      .then(doc => doc.toc)))
    .flatten()
    .map(asType(ACTIONS.tocLoaded))
    .replaceError(e => {
      return xs.of({
        type: ACTIONS.tocCacheError,
        data: e
      }); 
    });

  const fetchMissingTOC$ = action$
    .filter(action => {
      const { type, data: error } = action;
      return type === ACTIONS.tocCacheError && error.status === 404;
    })
    .map(fetchTOC)
    .flatten()
    .map(asType(ACTIONS.fetchedTOC));

  const saveFetchedTOC$ = action$.filter(withType(ACTIONS.fetchedTOC))
    .map(({ data: toc }) => xs.fromPromise(db.put({
      _id: "toc",
      toc
    })))
    .flatten()
    .map(toc => xs.of({
      type: ACTIONS.tocLoaded,
      data: toc
    }, {
      type: ACTIONS.savedTOC,
      data: toc
    }))
    .flatten();

  const fetchAllChapter$ = action$.filter(withType(ACTIONS.missingChaptersContent))
    .map(({ type, data: chapters }) => {
      /*
       * Array<Stream<Chapter>>
       */
      const chapters$Array = chapters.map((chapter, index) => xs
        .of(chapter)
        .compose(delay(1000 * index))
        .map(chapter => getContents$({
          author,
          repo,
          path: chapter.path
        })));

      return xs.merge.apply(null, chapters$Array).flatten();
    })
    .flatten()
    .map(asType(ACTIONS.chapterContentLoaded));

  // hwen tocs are loaded
  // look up in the db db for correspdongi chatpers
  const loadAllChapterFormCache$ = action$
    .filter(withType(ACTIONS.tocLoaded))
    .map(({ data: chapters }) => {

      /*
       * TODO: add test for this tiny tiny logic
       */
      return xs
        .fromPromise(db.bulkGet({
          docs: chapters.map(chapter => ({ id: chapter.path }))
        }))
        .map(response => {
          console.log(response);

          const {
            missing,
            found
          } = response.results
            .reduce((results, item) => {
              const OKs = item.docs
                .filter(({ ok = {} }) => {
                  return ok.chapter;
                } );

              if ( OKs.length > 0 ) {
                results.found.push(OKs[0].ok.chapter);
              } else {
                results.missing.push({
                  path: item.id
                });
              }
              return results;
            }, {
              missing: [],
              found: []
            });

          return xs.of(
            { 
              type: ACTIONS.loadedCachedChaptersContent,
              data: found
            },
            {
              type: ACTIONS.missingChaptersContent,
              data: missing
            }
          );
        })
        .flatten();

    })
    .flatten();

  // each time a chapter is loaded, save it in the local db
  const saveChapterContent$ = action$
    .filter(withType(ACTIONS.chapterContentLoaded))
    .map(({ data: chapter }) => xs
      .fromPromise(updateCachedChapter(chapter))
      .map(() => ({
        type: ACTIONS.chapterContentSaved,
        data: chapter
      })))
    .flatten();


  return xs.merge(
    loadTOC$,
    fetchMissingTOC$,
    saveFetchedTOC$,
    loadAllChapterFormCache$,
    fetchAllChapter$,
    saveChapterContent$,
    xs.of({ type: ACTIONS.loadTOC })
  )
    .replaceError(e => {
      console.error(e);
      return xs.empty();
    });
});
