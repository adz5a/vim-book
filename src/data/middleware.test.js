import { createMiddleware, toArray } from "xstream-redux-observable";
import { ACTIONS, withType, asType, makeReducer } from "./actions";
import xs from "xstream";
import flattenConcurrently from "xstream/extra/flattenConcurrently";
import delay from "xstream/extra/delay";
import PouchDB from "pouchdb-browser";
import { getContents } from "./api";
import { saveToc, loadToc, loadChapters } from "./middleware";


describe("pouchdb", () => {
  describe("get", () => {

    test("404", async function () {
      expect.hasAssertions();
      const db = new PouchDB("db-test");


      try {
        const response = await db.get("toc")
      } catch ( e ) {
        expect(e.status).toBe(404);
      }
      return await db.destroy();
    });

    test("200", async function () {
      expect.hasAssertions();
      const db = new PouchDB("db-test");
      const data = {
        foo: "bar"
      };

      await db.put({
        _id: "lol",
        data
      });


      try {
        const response = await db.get("lol");
        const { _id, _rev, data } = response;
        expect(_id).toBe("lol");
        expect(typeof _rev).toBe("string");
        expect(data).toEqual(data);
      } catch ( e ) {
        expect(e.status).toBe(404);
      }
      return await db.destroy();
    });

  });

  describe("put", () => {


    test("200", async function () {
      expect.hasAssertions();
      const db = new PouchDB("db-test");
      const data = {
        foo: "bar"
      };

      const response = await db.put({
        _id: "lol",
        data
      });

      const { ok, id, rev } = response;
      expect(ok).toBe(true);
      expect(id).toBe("lol");
      expect(typeof rev).toBe("string");

      return await db.destroy();
    });

    test("conflict", async function () {

      expect.hasAssertions();
      const db = new PouchDB("db-test");
      const data = {
        foo: "bar"
      };

      await db.put({
        _id: "lol",
        data
      });

      try {
        const response = await db.put({
          _id: "lol",
          data
        });
      } catch ( e ) {
        const { status, error, id } = e;
        expect({ status, error, id }).toEqual({
          status: 409,
          error: true,
          id: "lol"
        });
      }


      return await db.destroy();


    });

    test("update", async function () {

      expect.hasAssertions();
      const db = new PouchDB("db-test");
      const data = {
        foo: "bar"
      };

      const response = await db.put({
        _id: "lol",
        data
      });


      try {
        const final = await db.put({
          _id: response.id,
          _rev: response.rev,
          data: {
            foo: "baz"
          }
        });

        const { ok, rev, id } = final;
        expect(ok).toBe(true);
        expect(id).toBe("lol");
        expect(typeof rev).toBe("string");

      } catch ( e ) {
        console.log(e);
      }

      return await db.destroy();
    });

    describe("bulk add", () => {

      test("200", async function () {

        expect.hasAssertions();

        const messages = [ "bar", "baz", "boz" ].map(( m, i ) => ({
          _id: i + "_doc",
          data: {
            foo: m
          }
        }));

        const db = new PouchDB("db-test");

        try {
          const response = await db.bulkDocs(messages);

          response.forEach(({ ok, id, rev }) => {
            expect(ok).toBe(true);
            expect(typeof id).toBe("string");
            expect(typeof rev).toBe("string");
          });

        } catch ( e ) {
          console.log(e);
        }

        return await db.destroy();

      });

      test("update - with errors", async function () {

        expect.hasAssertions();

        const messages = [ "bar", "baz" ].map(( m, i ) => ({
          _id: i + "_doc",
          data: {
            foo: m
          }
        }));

        const db = new PouchDB("db-test");

        try {
          const response = await db.bulkDocs(messages);

          const updates = messages.map(( m, i ) => ({
            _id: response[i].id,
            _rev: i !== 1 ? "" : response[i].rev,
            data: {
              foo: "bar"
            }
          }));

          const final = await db.bulkDocs(updates);
          const [ r1, r2 ] = final;

          expect(r1.status).toEqual(409);
          expect(r1.error).toEqual(true);
          expect(r1.id).toEqual(messages[0]._id);

          expect(r2).toMatchObject({
            ok: true,
            id: messages[1]._id
          });

        } catch ( e ) {
          console.log(e);
        }

        return await db.destroy();

      });
    });
  });
});


describe("tocs", () => {

  const author= "sjl";
  const repo= "learnvimscriptthehardway";


  test("load toc - with empty db toc", async function () {
    expect.hasAssertions(); 

    const db = new PouchDB("toc-test");
    const res$ = await loadToc(db);

    const actions = await toArray(res$);
    // console.log(actions);

    const types = actions.map( a => a.type );
    expect(types).toMatchSnapshot();

    const toc = await db.get("toc");

    return await db.destroy();
  });

  test("load toc - with non-empty db toc", async function () {
    expect.hasAssertions(); 

    const db = new PouchDB("toc-test");
    await db.put({
      _id: "toc",
      toc: {
        foo: "bar"
      }
    });
    
    const res$ = await loadToc(db);
    const actions = await toArray(res$);
    // console.log(actions);

    const types = actions.map( a => a.type );
    expect(types).toMatchSnapshot();
    const [ action ] = actions;

    expect(action.data).toEqual({
      foo: "bar"
    });

    return await db.destroy();
  });


  test("load chapter", async function () {
    expect.hasAssertions();
    const db = new PouchDB("load-chapter-test");

    const toc = await getContents({
      author,
      repo,
      path: "chapters"
    });
    console.log("response");
    const res$ = loadChapters(toc.slice(0, 3));
    const final = await toArray(res$);

    const types = final.map( a => a.type )
      .forEach(t => expect(t).toBe(ACTIONS.chapterLoaded));

    return await db.destroy();

  });
});
