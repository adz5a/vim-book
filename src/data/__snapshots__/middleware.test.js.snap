// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tocs load toc - with empty db toc 1`] = `
Array [
  "toc-loaded",
  "toc-saved",
]
`;

exports[`tocs load toc - with non-empty db toc 1`] = `
Array [
  "toc-loaded",
]
`;
